dot/parachain/dispute/backend.go

```diff
@@ -0,0 +1,207 @@
+package dispute
+
+import (
+	"sync"
+	"time"
+
+	"github.com/ChainSafe/gossamer/dot/parachain/dispute/types"
+	parachainTypes "github.com/ChainSafe/gossamer/dot/parachain/types"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/dgraph-io/badger/v4"
+	"github.com/tidwall/btree"
+)
+
+// Backend is the backend for the dispute coordinator module.
+type Backend interface {
+	// GetEarliestSession returns the earliest session index, if any.
+	GetEarliestSession() (*parachainTypes.SessionIndex, error)
+	// GetRecentDisputes returns the recent disputes, if any.
+	GetRecentDisputes() (*btree.BTree, error)
+	// GetCandidateVotes returns the votes for the given candidate for the specific session-candidate pair, if any.
+	GetCandidateVotes(session parachainTypes.SessionIndex, candidateHash common.Hash) (*types.CandidateVotes, error)
+
+	// SetEarliestSession sets the earliest session index.
+	SetEarliestSession(session *parachainTypes.SessionIndex) error
+	// SetRecentDisputes sets the recent disputes.
+	SetRecentDisputes(recentDisputes *btree.BTree) error
+	// SetCandidateVotes sets the votes for the given candidate for the specific session-candidate pair.
+	SetCandidateVotes(session parachainTypes.SessionIndex, candidateHash common.Hash, votes *types.CandidateVotes) error
+}
+
+// OverlayBackend is the overlay backend for the dispute coordinator module.
+type OverlayBackend interface {
+	Backend
+
+	// WriteToDB writes the given dispute to the database.
+	WriteToDB() error
+	// GetActiveDisputes returns the active disputes.
+	GetActiveDisputes(now int64) (*btree.BTree, error)
+}
+
+// DBBackend is the backend for the dispute coordinator module that uses a database.
+type DBBackend interface {
+	Backend
+
+	// Write writes the given data to the database.
+	Write(earliestSession *parachainTypes.SessionIndex,
+		recentDisputes *btree.BTree,
+		candidateVotes map[types.Comparator]*types.CandidateVotes) error
+}
+
+type syncedEarliestSession struct {
+	*sync.RWMutex
```

Comment:
Why is the mutex a pointer?

Reply:
Hmm, I don't recall. It could have been cause of a Deepsource warning, I'm not sure.

Reply:
If its just for deepsource and the issue is embedding the mutex, perhaps it's better to have it as an attribute rather than embed?

Reply:
No, the issue was something to do with not using the pointer. Could you help me understand what's wrong with the embed pointer?

Reply:
I'll also check the deepsource and get more data on the warning!

Reply:
I dont have any issue with it, I just thought that I recalled getting deepsource issues about embedding mutexs, but if this is not an issue sweet! Aside from that was just curious why a pointer was preferred over non-pointer, but I have no objection to its use

Reply:
and deepsource doesn't complain about it anymore, haha. I've removed the pointer embed